File: project_toutiao/crud/news.py
from sqlalchemy import func, select, update
from sqlalchemy.ext.asyncio import AsyncSession

from models.news import Category, News


async def get_categories(db: AsyncSession, skip: int = 0, limit: int = 100):
    result = await db.execute(select(Category).offset(skip).limit(limit))
    return result.scalars().all()


async def get_news_list(
    db: AsyncSession, category_id: int, skip: int = 0, limit: int = 100
):
    stmt = (
        select(News)
        .where(News.category_id == category_id)
        .offset(skip)
        .limit(limit)
    )
    result = await db.execute(stmt)
    return result.scalars().all()


async def get_news_count(db: AsyncSession, category_id: int):
    stmt = select(func.count(News.id)).where(News.category_id == category_id)
    result = await db.execute(stmt)
    return result.scalar_one()


async def get_news_detail(db: AsyncSession, news_id: int):
    result = await db.execute(select(News).where(News.id == news_id))
    return result.scalar_one_or_none()


async def increase_news_views(db: AsyncSession, news_id: int):
    stmt = update(News).where(News.id == news_id).values(views=News.views + 1)
    result = await db.execute(stmt)
    await db.commit()
    return result.rowcount > 0


async def related_news(
    db: AsyncSession, category_id: int, news_id: int, limit: int = 5
):
    stmt = (
        select(News)
        .where(News.id != news_id, News.category_id == category_id)
        .order_by(News.views.desc(), News.publish_time.desc())
        .limit(limit)
    )
    result = await db.execute(stmt)
    rows = result.scalars().all()
    return [
        {
            "id": n.id,
            "title": n.title,
            "content": n.content,
            "image": n.image,
            "author": n.author,
            "publishTime": n.publish_time,
            "category": n.category_id,
            "views": n.views,
        }
        for n in rows
    ]

File: project_toutiao/frontend/src/utils/news.ts
import type { NewsSummary } from "@/types/news";

export function normalizeNewsSummary(raw: Record<string, unknown>): NewsSummary {
  const id = Number(raw.id);
  const categoryId = Number(raw.categoryId ?? raw.category_id ?? 0);
  const publishTime =
    (raw.publishTime ?? raw.publish_time) as string | null | undefined;
  return {
    id,
    title: String(raw.title ?? ""),
    description: (raw.description as string | null) ?? null,
    image: (raw.image as string | null) ?? null,
    author: (raw.author as string | null) ?? null,
    publishTime: publishTime ?? null,
    categoryId,
    views: Number(raw.views ?? 0),
  };
}
